how-it-works/1_protocol/04_execution.subpage.md

```diff
@@ -5,8 +5,92 @@ shareImage: /img/how-it-works/network-nervous-system-nns.600.jpg
 slug: execution-layer
 ---
 
-The execution layer processes one input at a time. This input is taken from one of the
-input queues, and is directed to one canister. Based on this input and the state of the
-canister, the execution environment updates the state of the canister, and additionally may
-add messages to output queues and update the ingress history (possibly with a response to
-an earlier ingress message).
\ No newline at end of file
+# Execution
+
+The execution layer is the topmost layer of the IC core protocol stack.
+It deterministically schedules and executes the messages that have been agreed on by consensus and inducted into the canister queues by message routing, thereby changing the state of the subnet in a deterministic manner on all the nodes.
+The execution of the same sequence of messages on every node of the subnet guarantees that the same ending state is obtained on each node of the subnet after completion of the round. 
+
+A canister smart contract on the IC is the union of Web Assembly (Wasm) bytecode representing the smart contract program and memory pages representing its state.
+The Wasm bytecode can be modified by installing or updating the canister.
+The smart contract state gets modified when executing messages on the canister smart contract.
+Both the bytecode and the memory pages, i.e., the state, of the canister, are maintained by every node machine of the subnet the canister is installed on.
+Each node in the subnet holding the same canister state and ensuring that the state transitions in the same way on every node in every round is the foundation of realizing a replicated state machine and the security and resilience properties thereof that make blockchains so unique.
+
+## Concurrent Message Execution
+
+In one IC round, the message routing layer invokes the execution layer once for executing (a subset of) the messages in the canister input queues.
```

Comment:
I think we should introduce "update calls" in this section rather than leave it to the non-replicated execution section.

Reply:
Noted for iteration 2!